ignore/src/walk.rs

```diff
@@ -2071,4 +2075,42 @@ mod tests {
         builder.follow_links(true).same_file_system(true);
         assert_paths(td.path(), &builder, &["same_file", "same_file/alink"]);
     }
+
+    #[cfg(target_os = "linux")]
+    fn chmod<P: AsRef<Path>>(path: P, mode: libc::mode_t) -> Result<(), std::io::Error> {
+        use std::os::unix::ffi::OsStringExt;
+        let bytes = path.as_ref().as_os_str().to_os_string().into_vec();
+        let cstring = std::ffi::CString::new(bytes).unwrap();
+        let ret = unsafe { libc::chmod(cstring.as_ptr(), mode) };
+        if ret < 0 {
+            Err(std::io::Error::last_os_error())
+        } else {
+            Ok(())
+        }
+    }
+
+    #[test]
+    #[cfg(target_os = "linux")]
+    fn no_read_permissions() {
+        struct ChmodOnDrop(std::path::PathBuf, libc::mode_t);
+
+        impl Drop for ChmodOnDrop {
+            fn drop(&mut self) {
+                chmod(&self.0, self.1).unwrap();
+            }
+        }
+
+        let td = tmpdir();
+        let dir_path = td.path().join("a");
+        mkdirp(&dir_path);
+        wfile_size(dir_path.join("foo"), 0);
+
+        {
+            chmod(&dir_path, 0).unwrap();
+            let _chmod_on_drop = ChmodOnDrop(dir_path, libc::S_IRUSR | libc::S_IWUSR | libc::S_IXUSR);
+
+            let builder = WalkBuilder::new(td.path());
+            assert_paths(td.path(), &builder, &["a"]);
+        }
+    }
```

Comment:
Hmm, yeah, tests should not fail depending on what user ran them. Maybe it would be good to check if the file is readable after chmodding it. If it is, then skip the test.

With that said, it would be nice if we could avoid bringing in `libc` and `unsafe` for this test. Is there some other way to test this? Nothing is coming to mind... Maybe using a pre-existing file that a normal user traditionally doesn't have access to? (And skipping the test if the file doesn't exist or is readable.)

Reply:
Please also make sure that lines are wrapped to 79 columns, inclusive.

Reply:
@BurntSushi, I added the proposed check, though, of course, it'd be better if the test harness allowed marking a test as skipped at runtime.

With regard to avoiding adding a dev-dependency on `libc` I don't have a good idea either. But now I made it a Linux-only dependency. Perhaps that's good enough?

If using `libc` is really an issue (even though it already is a regular dependency for some other ripgrep crates, though, in truth, not ignore itself), then I'll happily change this to try to walk /root and check if it is observing at least one valid entry.

Reply:
I probably care more about the extra `unsafe` issue to be honest. I'm mostly just trying to make sure things are as a tight as possible.

Just looking at my file system, maybe just trying look for `/etc/sudoers.d`? It's readable by `root`, but not by anyone else. So I think the logic might be something like this:

* Stat `/etc/sudoers.d` via `std::fs::metadata()`.
* If it errors, skip test.
* Read `/etc/sudoers.d`.
* If it succeeds, skip test.
* Try to traverse `/etc/sudoers.d` and run the test normally.

Reply:
@BurntSushi, alright, sounds like a plan! I pushed the proposed change. Without a doubt that made the test much shorter and simpler. Thanks for the review!

Reply:
Maybe `/root` (or more generally `~root`) is more ubiquitous than `/etc/sudoers.d`?